hal/test-headers/ticker_api_tests.h

```diff
@@ -0,0 +1,138 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2017-2017 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/** \addtogroup hal_specification */
+/** @{*/
+
+#ifndef MBED_US_TICKER_API_TESTS_H
+#define MBED_US_TICKER_API_TESTS_H
+
+#include "device.h"
+
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * \defgroup hal_specification_ticker Ticker
+ * Ticker specification and tests
+ *
+ * To run the microsecond ticker hal tests use the command:
+ *
+ *     mbed test -t <toolchain> -m <target> -n tests-mbed_hal-us_lp_ticker*
+ *
+ * # Defined behavior
+ * * The function ticker_init is safe to call repeatedly - Verified by test ::ticker_init_test
+ * * The function ticker_init resets the internal count and disables the ticker interrupt - Verified by test ::ticker_init_test
+ * * Ticker frequency is non-zero and counter is at least 8 bits - Verified by ::ticker_info_test
+ * * The ticker rolls over at (1 << bits) and continues counting starting from 0 - Verified by ::ticker_overflow_test
+ * * The ticker counts at the specified frequency +- 10% - Verified by ::ticker_frequency_test
```

Comment:
We could test this by counting ticker ticks in the specified time period and calculating clock frequency, but there is a problem with time pattern. We can not use wais_us() since it also uses ticker us_ticker_read() and additionally it does not take into account ticker frequency (it assumes that ticker clock speed is 1MHz => 1 tick == 1 us). We might use RTC and count seconds, but RTC accuracy is poor. What do you suggest?

Reply:
It should probably be compared against the clock of the PC running the test. This is already done with some timing tests such as ```mbed-os\TESTS\mbed_drivers\ticker\main.cpp```.

Reply:
Thanks for the answers. Still it would be also good to provide fix for wait_us(). wait_us() should use ticker_read_us(const ticker_data_t *const) instead of us_ticker_read/hf_ticker_read. I can provide this fix with the tests and update of ticker IRQ handler.
Additionally how the previous version worked on the boards where ticker clock was not 1MHz?